### PR TITLE
Player

### PR DESCRIPTION
**Pull request by BenGriffith**

### CODE REVIEW

File: poker/utils/action.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+from poker.utils.chip import GameStack
+
+
+class Action:
+
+    CHECK = "check"
```
Comment (bbelderbos) [inline poker/utils/action.py:6]: What about grouping them in an enum like you did with FaceCards?

File: poker/utils/action.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    RAISE = "raise"
+    CALL = "call"
+    FOLD = "fold"
+
+    def __init__(self, game_stack: GameStack, player: any = None) -> None:
+        self.game_stack = game_stack
+        self.person = player
+
+    def check(self) -> None:
```
Comment (bbelderbos) [inline poker/utils/action.py:15]: Still to be implemented?

File: poker/utils/chip.py [hunk 3]
```diff
@@ -1,54 +1,53 @@
...
-class Stack:
+class PlayerStack:
+
+    WHITE = {"name": "White", "value": 1}
+    RED = {"name": "Red", "value": 5}
+    BLUE = {"name": "Blue", "value": 10}
 
     def __init__(self) -> None:
-        self.chips = defaultdict(int)
+        self.chips = Counter()
```
Comment (bbelderbos) [inline poker/utils/chip.py:14]: great, more suitable data structure

File: poker/utils/chip.py [hunk 4]
```diff
@@ -1,54 +1,53 @@
...
-            self.chips[Chip.WHITE.name] += SINGLE_CHIP * Cash.FIVE.value
+            self.chips[self.WHITE["name"]] += SINGLE_CHIP * Cash.FIVE.value
         elif cash == Cash.TEN.value:
-            self.chips[Chip.WHITE.name] += SINGLE_CHIP * Cash.FIVE.value
-            self.chips[Chip.RED.name] += SINGLE_CHIP
+            self.chips[self.WHITE["name"]] += SINGLE_CHIP * Cash.TEN.value
         elif cash == Cash.FIFTEEN.value:
-            self.chips[Chip.RED.name] += SINGLE_CHIP
-            self.chips[Chip.BLUE.name] += SINGLE_CHIP
+            self.chips[self.WHITE["name"]] += SINGLE_CHIP * Cash.FIFTEEN.value
```
Comment (bbelderbos) [inline poker/utils/chip.py:29]: If it's a self.attr then it's not a constant right? So better lowercase them?

File: poker/utils/chip.py [hunk 5]
```diff
@@ -1,54 +1,53 @@
...
+class GameStack(PlayerStack):
 
-    # count
-    def chip_count(self) -> tuple:
-        return (
-            self.chips[Chip.WHITE.name],
-            self.chips[Chip.RED.name],
-            self.chips[Chip.BLUE.name],
-            )
+    def __init__(self) -> None:
```
Comment (bbelderbos) [inline poker/utils/chip.py:49]: Curious if you need this if nothing gets done in constructor?

File: poker/utils/constants.py [hunk 6]
```diff
@@ -1,22 +1,40 @@
...
 
-SUITS = ["hearts", "diamonds", "spades", "clubs"]
-FACE_CARDS = "J Q K A".split()
+SUITS = f"{emoji.emojize(':heart_suit:')} {emoji.emojize(':diamond_suit:')} {emoji.emojize(':spade_suit:')} {emoji.emojize(':club_suit:')}".split()
 NUMBER_CARDS = list(range(2, 11))
-INCREMENT_LIMIT = 20
+INCREMENT_LIMIT = 100
 SINGLE_CHIP = 1
 DOUBLE = 2
+COMPETITION = (list(range(1, 5)))
```
Comment (bbelderbos) [inline poker/utils/constants.py:10]: why wrap in `()` here?

File: poker/utils/constants.py [hunk 7]
```diff
@@ -1,22 +1,40 @@
 from enum import Enum
 
+import emoji
 
-SUITS = ["hearts", "diamonds", "spades", "clubs"]
-FACE_CARDS = "J Q K A".split()
+SUITS = f"{emoji.emojize(':heart_suit:')} {emoji.emojize(':diamond_suit:')} {emoji.emojize(':spade_suit:')} {emoji.emojize(':club_suit:')}".split()
 NUMBER_CARDS = list(range(2, 11))
```
Comment (bbelderbos) [inline poker/utils/constants.py:6]: might not need the list() here?

File: poker/utils/constants.py [hunk 8]
```diff
@@ -1,22 +1,40 @@
...
+    A = 14
 
 
 class Cash(Enum):
     FIVE = 5
     TEN = 10
     FIFTEEN = 15
-    TWENTY = 20
\ No newline at end of file
+    TWENTY = 20
```
Comment (bbelderbos) [inline poker/utils/constants.py:26]: nice grouping of constants in module

File: poker/utils/constants.py [hunk 9]
```diff
@@ -1,22 +1,40 @@
...
+    FIFTY = 50
+    HUNDRED = 100
+
+
+class Blind(Enum):
```
Comment (bbelderbos) [inline poker/utils/constants.py:31]: IntEnum?

gpt:
> An IntEnum is a subclass of int that provides the same benefits of an Enum (clear and self-documenting values, prevention of magic numbers and string literals), but also allows for arithmetic operations and comparisons between its values.

So I don't think it matters here actually

File: poker/utils/exception.py [hunk 10]
```diff
@@ -12,4 +12,37 @@ class CashException(BaseException):
...
+class NotReadyException(BaseException):
+    pass
+
+
+class InvalidActionException(BaseException):
+    pass
+
+
+class InsufficientChipException(BaseException):
+    pass
```
Comment (bbelderbos) [inline poker/utils/exception.py:48]: Nice use of own exception, naming is clear, if you want to provide more info when to use you could replace "pass" by a docstring

File: poker/utils/message.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        if player_response >= chip_count:
+            raise InsufficientChipException
+        return player_response
+    
+    def different_amount(self, chip_count: int) -> None:
+        print(f"You only have {chip_count} chips. Please select a lower amount.")
+    
+    def player_summary(self, players: dict) -> None:
+        player_table = Table(title="Player Summary")
+        player_table.add_column("Order")
```
Comment (bbelderbos) [inline poker/utils/message.py:87]: could do these in a loop maybe? for col in col_list: add_column

File: poker/utils/message.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        player_table.add_column("Blind")
+        player_table.add_column("Pocket Cards")
+        player_table.add_column("Best Hand")
+        player_table.add_column("Short Name")
+        for player_id, player in players.items():
+            player = player["player"]
+            player_table.add_row(
+                str(player_id), 
+                player.name, 
+                f"{GameStack.WHITE['name']}: {player.stack.chips[GameStack.WHITE['name']]}",
```
Comment (bbelderbos) [inline poker/utils/message.py:99]: this is a bit hard to read, can you move all this inline code into intermediate variables so the add_row becomes more concise?

File: poker/utils/message.py [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+        try:
+            if progress:
+                player_response = input(f"How about now, are you ready? [yes/no] ").lower()
+            else:
+                player_response = input(f"Are you ready to continue? [yes/no] ").lower()
+            if player_response not in self.decision:
+                raise GamePlayException
+            if player_response in [Decision.N.value, Decision.NO.value]:
+                raise NotReadyException
+        except GamePlayException:
```
Comment (bbelderbos) [inline poker/utils/message.py:118]: I wonder if raise + catch in same place is kind of redundant, can you just do 

```
if player_response not in self.decision:
    # code that happens under "except ..."
```

File: poker/utils/message.py [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+        console.print(Columns(game_pot))
+
+    def showdown(self, winner: dict, pot: GameStack, players: dict, community_cards: list[Card]) -> None:
+        sleep(GAME_DELAY)
+        self.game_summary(pot=pot, community_cards=community_cards)
+        sleep(GAME_DELAY)
+        self.player_summary(players=players)
+        sleep(GAME_DELAY)
+
+        if len(winner) > 3:
```
Comment (bbelderbos) [inline poker/utils/message.py:138]: Ideally you put magic numbers into constants as well

File: poker/utils/message.py [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+            winners = len(winner) // 2
+            winnings = pot.cash_equivalent() / winners
+            winners_names = [value for key, value in winner.items() if "name" in key]
+            winners_names_string = ", ".join(winners_names)
+            print(f"\nCongratulations {winners_names_string}! You each won ${winnings}!")
+        if len(winner) == 3:
+            if winner["name"] == "You":
```
Comment (bbelderbos) [inline poker/utils/message.py:145]: ditto for these kind of strings

File: poker/utils/player.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
+import random
+
+from poker.utils.chip import PlayerStack
+from poker.utils.deck import Deck, Card
+from poker.utils.action import Action
+
+
+class Player:
+
+    PLAYER = "Player"
```
Comment (bbelderbos) [inline poker/utils/player.py:10]: enum?

Comment (bbelderbos) [reply poker/utils/player.py:10]: like you did with Action.Constant below?

File: tests/test_message.py [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+    ("FOLD", [Action.CALL, Action.FOLD]),
+    ("fold", [Action.CALL, Action.FOLD])
+])
+def test_action_valid_response_raise(monkeypatch, action_raise_prompt, test_input, expected):
+    monkeypatch.setattr("builtins.input", lambda _: test_input)
+    response = input(action_raise_prompt).lower()
+    assert response in expected
+
+
+@pytest.mark.parametrize("test_input, expected", [
```
Comment (bbelderbos) [inline tests/test_message.py:115]: TODO: need to look at parametrize in this module.

**Review by bbelderbos — COMMENTED** [review]
First code review pass, HTH

File: poker/utils/action.py (hunk 1 above)
Comment (BenGriffith) [reply poker/utils/action.py:6]: I'll go that route. I wasn't certain about making class attributes to begin with.

File: poker/utils/action.py (hunk 2 above)
Comment (BenGriffith) [reply poker/utils/action.py:15]: I removed this method. It will be removed in a future PR.

File: poker/utils/chip.py (hunk 4 above)
Comment (BenGriffith) [reply poker/utils/chip.py:29]: Thanks for the guidance here.

File: poker/utils/message.py (hunk 13 above)
Comment (BenGriffith) [reply poker/utils/message.py:118]: Good call. It is redundant. Your suggestion works much better.

File: poker/utils/chip.py (hunk 5 above)
Comment (BenGriffith) [reply poker/utils/chip.py:49]: I think this is needed. I ran a quick test by removing it and program threw an error.

File: poker/utils/message.py (hunk 14 above)
Comment (BenGriffith) [reply poker/utils/message.py:138]: ah, I overlooked both of these. Thanks.

File: poker/utils/player.py (hunk 16 above)
Comment (BenGriffith) [reply poker/utils/player.py:10]: You got it. I'll work on this update. Thanks.

File: poker/utils/constants.py (hunk 6 above)
Comment (BenGriffith) [reply poker/utils/constants.py:10]: Good catch. I'm not sure why I did that. I'll remove it.

File: poker/utils/constants.py (hunk 7 above)
Comment (BenGriffith) [reply poker/utils/constants.py:6]: I removed the list and didn't work. I suppose I could remove it here and then update the deck module.

File: poker/utils/message.py (hunk 11 above)
Comment (BenGriffith) [reply poker/utils/message.py:87]: good call. I'll rework it.

File: poker/utils/message.py (hunk 12 above)
Comment (BenGriffith) [reply poker/utils/message.py:99]: completely agree. I'll work on making it more readable/concise.

**Pull request merged**